묵찌빠 게임 [Step1] Morgan, Kiseok

devjoon opened the pull request:
@hoon94
안녕하세요 hoon ! 2조 Morgan , Kiseok입니다 첫번째 PR보내 드립니다. 
피드백 주시는 문제에 대해서 열심히 수정해 보겠습니다!

## 고민되었던 점
- (코드41번줄) 처음 userHand에 기본값을 .rock으로 넣어줬는데 기본값을 넣어주지않으면 에러가 나서 nil처럼 초기값을 주지않고 나중에 처리 할 수 있는 방법이 있을지 궁금합니다.<br> -> rockpaperscissors 열거형에 아무것도 아닌 상태의 case 정의

- 가위바위보를 enum타입으로 정의하고 인스턴스를 만들어 준 후 switch 문을 사용하여 숫자에 맞춰 가위,바위,보를 할당해주었는데 함수가 너무 길어지게 되었습니다. rawValue를 사용해보려 하였으나 코드에서 승패 결과 조건이 직관적으로 보이지 않아 사용하지 못하였습니다.<br> -> 사용자와 컴퓨터의 가위,바위,보를 판단하는 함수 분리

## 조언을 얻고 싶은 부분
- 저희가 구현한 열거형이 옳게 사용된 것인지 궁금합니다.
- 효율성을 좀 더 올릴수 있는 부분이 있다면 알고싶습니다.

devjoon committed:
feat: 결과,에러 에 대한 열거형 정의, 가위바위보 게임 구조체 정의

carti1108 committed:
feat: 사용자의 입력값과 컴퓨터의 입력값을 비교하여 결과를 출력해주는 함수 작성

devjoon committed:
feat: 에러핸들링 및 실행예시 구현함수 작성

carti1108 committed:
refactor: 함수 분리

readUserInput함수 기능 분리

devjoon committed:
Update README.md

devjoon committed:
fix: 예외처리 수정

게임 종료시 (0 입력시) 오출력되는 부분 수정
잘못된 입력 처리후 게임 진행시 게임결과 부분 오출력되는 오류 수정

Hoon94 commented on RockPaperScissors/RockPaperScissors/enumType.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+}
+
+enum InputError: LocalizedError {
```
`LocalizedError`를 사용하여 에러에 대한 처리를 해주신 부분이 대단합니다.👍

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+    guard let user = Int(userInput) else {
+        throw InputError.wrongInput
+    }
+    
```
위에서와는 다르게 여기서는 개행이 추가되었네요. 두 분이서 정리하신 하나의 컨벤션으로 통일해 주시면 좋을 것 같습니다.😉
아직 정해진 컨벤션이 없다면 [공식 문서](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/thebasics/)의 컨벤션을 따라도 좋을 것 같습니다.

또는 아래와 같이 대표적인 컨벤션도 있으니 통일성 있게 사용하시면 좋을 것 같습니다.
- [StyleShare](https://github.com/StyleShare/swift-style-guide)
- [Swift Style Guide](https://google.github.io/swift/#column-limit)

devjoon replied:
위에는 guard let 바인딩이고 아래쪽은 guard문으로 조건 확인용도로 사용하여 저희는 개행추가하는 판단을 하였습니다.
피드백을 토대로 guard문 사용하면서 사이에 개행은 제거하도록 하겠습니다 감사합니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
-import Foundation
+var userHand: RockPaperScissors = .nothing
+var computerHand: RockPaperScissors = .nothing
+var gameResult: Result = .draw
```
전역 변수를 사용하셨을 때 장점과 단점은 무엇일까요?🧐

carti1108 replied:
전역변수 사용 시 전역에서 사용이 가능하고 값을 저장했을 때 지역변수와는 다르게 값이 사라지지 않습니다.
그렇기 때문에 메모리 측면에서 낭비가 생기고 프로젝트가 커질 수록 가독성이 떨어질 수 있습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+    }
+    
+    guard 0...3 ~= user else {
+        throw InputError.wrongInput
```
열거형으로 만든 `InputError`를 조금 더 세분화하여 각각이 어떤 에러인지를 조금 더 명확하게 구분해 주시면 좋을 것 같습니다. 지금처럼 코드가 작으면 어디서 발생한 에러인지 찾기 쉽지만 큰 프로젝트에서는 어디서 발생한 에러인지 확인하기 어려울 것 같습니다.😅

carti1108 replied:
말씀해주신대로 에러 케이스들은 세분화 하였습니다. 하지만 프로젝트 실행예시에서는 어떤 오류든 "잘못된 입력입니다. 다시 시도해주세요"만 출력하기 때문에 localized error는 빼고 Error 프로토콜을 따르도록 하였습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+        userHand = .rock
+    case 3:
+        userHand = .paper
+    default:
```
여기서는 `default`를 작성하여 `switch`문을 사용하셨네요. `enum`을 사용할 때와는 어떤 차이가 있을까요?

carti1108 replied:
user를 enum타입으로 정의 한다면 switch문처럼 default를 이용하여 예외처리를 할 필요없이 enum타입의 case만으로 처리할 수 있다고 생각합니다. 저희는 이미 만들어 놓은 RockPaperScissors enum을 활용하여 switch문을 통해 user를 만든다면 새로운 user enum타입을 정의할 수고를 덜수 있다고 생각하였습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+}
+
+func makeComputerHand() {
+    let selectComputerHand = Int.random(in: 1...3)
```
1. `selectComputerHand`라는 네이밍에서 `select`는 동사입니다. 상수에 맞게 수정해 보면 어떨까요?
2. `random` 함수 외에 랜덤 한 값을 생성할 수 있는 방법에는 또 어떤 다른 방식이 있을까요? 어떤 것들이 있는지 찾아보고 `random` 함수를 사용하면 어떤 장점이 있는지 알려주시면 좋겠습니다.🤩
3. `magic literal`은 무엇일까요? 사용하신 [`magic literal`](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)을 어떻게 수정할 수 있을까요?

carti1108 replied:
1. select에서 random으로 수정하였습니다.
2. array에서 randomElemente() 함수를 사용하여 추출하는 방법도 있습니다. random함수 사용 시 randomElemente함수와 다르게 옵셔널 바인딩을 할 필요가 없으나 Int만 반환할 수 있다는 단점이 있습니다.
https://developer.apple.com/documentation/swift/array/randomelement()
3. 매직 리터럴은 프로그래밍 상 사용되는 숫자나 문자를 그대로 사용하는 것이라고 알고있습니다. 이 부분은 상수에 할당을 해주어 변경하였습니다.

Hoon94 replied:
시간 복잡도란 무엇일까요? `random`과 `randomElement`의 시간 복잡도는 각각 어떤가요?

devjoon replied:
시간 복잡도란 해당 함수를 실행함에 있어 걸리는 시간을 빅오 표기법으로 표현한것을 말 합니다

randomElement 의 경우 배열에 있는 원소중 랜덤으로 가져와야하기때문에 전체적으로 탐색을 해야해서 선형 탐색을 이용해야할것 같습니다 그 경우 빅오 표기법으로는 O(N) 으로 표기될수 있습니다

random같은 경우는 무작위로 범위내의 숫자가 생성되기 때문에 빅오 표기법으로는 O(1) 으로 표기될 것 같습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+        computerHand = .rock
+    case 3:
+        computerHand = .paper
+    default:
```
위의 코멘트처럼 여기서도 `default`를 작성하여 `switch`문을 사용하셨네요.

carti1108 replied:
위에 했던 답변과 동일합니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+    if computerHand == userHand {
+        gameResult = .draw
+        print(gameResult.result)
+        playGame()
```
재귀 함수를 사용하여 코드를 구현하셨네요.👍
재귀 함수를 사용하지 않고 구현할 수 있는 방법은 어떤 게 있을까요?
다른 방법과 비교하여 각각의 장단점은 무엇일까요?

devjoon replied:
for 문이나 while문을 이용하여 처리할수 있을 것 같습니다
재귀함수를 사용할경우 코드가 반복문보다 간결해진다는 장점이 있습니다
단점으로는 스택오버플로우가 발생하여 성능의 저하가 발생할수 있습니다

Hoon94 commented on RockPaperScissors/RockPaperScissors/enumType.swift:
```diff
@@ -0,0 +1,40 @@
+import Foundation
+
+enum RockPaperScissors {
+    case rock, paper, scissors, nothing
```
`RockPaperScissors`에서 `nothing`은 어떤 의미가 있을까요? 현재 코드에서는 사용자 입력이 0일 때 게임 종료를 위한 `case`로 보입니다. 조금 더 직관적인 네이밍은 어떨까요?😄

devjoon replied:
nothing -> exit 로 좀 더 직관적으로 변경하였습니다 감사합니다.

carti1108 replied:
처음에는 인스턴스 생성 시 초기화를 해주어야해서 마땅히 초기값으로 넣어줄 값이 없어서 만든 부분이었습니다. 하지만 exit 이란 표현을 사용하면 될 것 같아 exit case를 추가해주었습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+    }
+}
+
+func makeResult() {
```
`make`를 사용하여 여러 함수를 네이밍 하셨습니다. [API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)에서는 다음과 같은 사용을 권장하고 있습니다.🤗
```
Begin names of factory methods with “make”, e.g. x.makeIterator().
```
다른 네이밍으로 수정해 보는 건 어떠실까요?

carti1108 replied:
make를 사용한 함수들을 generate로 수정하였습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/enumType.swift:
```diff
@@ -0,0 +1,40 @@
+    case rock, paper, scissors, nothing
+}
+
+enum Result {
```
[Result](https://developer.apple.com/documentation/swift/result)는 이미 사용되고 있는 타입명입니다. 다른 네이밍으로 수정해 보는 건 어떨까요?

carti1108 replied:
Result를 GameResult로 변경 해주었습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,86 @@
+        throw InputError.wrongInput
+    }
+    
+    try makeUserHand(user: user)
```
`makeUserHand(user:)` 함수가 `readUserInput()` 함수에서 호출됩니다. `readUserInput()` 함수를 직역해 보면 사용자의 입력을 받아오는 함수로 생각됩니다.
`makeUserHand(user:)` 함수가 이 안에 위치하는 것이 좋을까요?🤔

carti1108 replied:
generateUserHand함수를 생성하여 그 안에서 readUserInput과 makeUserHand함수를 호출하도록 변경 하였습니다.
makeUserHand의 함수명은 convertToUserHand로 변경 하였습니다.

Hoon94 commented on README.md:
```diff
@@ -1,6 +1,13 @@
-### 묵찌빠 프로젝트 저장소
-
-- 이 저장소를 자신의 저장소로 fork하여 프로젝트를 진행합니다
+# ✊ 묵찌빠 게임
```
`README` 벌써 작성하시다니 멋진 습관입니다.🤩 또한 `Flow Chart` 또한 잘 작성되어 있는 것 같습니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/enumType.swift:
```diff
@@ -0,0 +1,40 @@
+enum Result {
+    case win, draw, lose, exit
+
+    var result: String {
```
연산 프로퍼티를 사용하여 가독성이 뛰어납니다.👍

Hoon94 reviewed (commented):
안녕하세요. Kiseok, Morgan👋
요구 조건에 맞춰 스텝을 진행하시느라 고생 많으셨습니다. 일주일 동안 잘 부탁드립니다.😆
코멘트 읽어 보시고 설명이 부족하거나 이해가 안 가시는 부분이 있다면 언제든 편하게 질문 남겨주세요.🙌
또한 어디까지나 정답이 아닌 의견을 말씀드린 부분이라는 점을 기억하시고 다른 의견이 있으시다면 주저 없이 말씀해 주세요.🙏

## 칭찬 드리고 싶은 부분

- 각각의 함수에 대해 기능을 분리하려고 노력하신 점👍
- 함수의 배치를 신경 써서 가독성을 높이신 점👍
- 각각의 커밋에 대해 일관된 컨벤션을 유지하신 점👍

## 조언을 얻고 싶은 부분

> Q: 저희가 구현한 열거형이 옳게 사용된 것인지 궁금합니다.

A: 열거형을 사용하여 케이스에 대한 값을 나누어 깔끔한 코드를 작성하신 부분이 대단합니다. 아래 추가적인 코멘트를 통해 열거형을 조금 수정해 보시면 더 좋을 것 같습니다.😊

> Q: 효율성을 좀 더 올릴수 있는 부분이 있다면 알고싶습니다.

A: 코드를 작성하며 어떤 식으로 효율성을 올릴지에 대한 고민은 항상 필요한 것 같습니다.🥺 이번 스텝에서는 다음과 같은 키워드로 추가 작업을 진행해 보시면 좋을 것 같습니다. 

- `시간 복잡도`, `재귀 함수와 반복문`

## Tip

1. 커밋 내역을 조금 더 세분화하면 좋을 것 같습니다. 
예를 들어 첫 커밋인 `feat: 결과,에러 에 대한 열거형 정의, 가위바위보 게임 구조체 정의`의 경우에서도 열거형 정의와 구조체 정의를 구분하여 커밋하면 좋을 것 같습니다.
2. 이번 미션부터는 타입을 사용할 수 있는 것으로 알고 있습니다. `main`에서 사용하고 있는 변수와 함수를 타입을 사용하여 수정해 보면 어떨까요?
3. 미션을 진행하시면서 STEP 단위로 `git branch`를 생성하는 것도 좋은 방법일 것 같습니다.
4. 스위프트에서는 어떤 식으로 파일을 구분하여 관리할지 고민해 보시면 좋을 것 같습니다. [Swift github](https://github.com/apple/swift/tree/main/stdlib/public/core)와 다른 예들을 찾아보면 좋을 것 같습니다.

devjoon committed:
refactor: 에러케이스 세분화, 함수명 변경

에러케이스를 각 조건에 맞춘 에러들로 세분화
make가 사용된 함수명을 generated로 변경
enum RockPaperScissors 의 .nothing을 .exit로 변환
selectComputerHand 상수를 randomComputerHand로 변환

carti1108 committed:
refactor: 함수 위치 변경 및 함수명 변경

generateUserHand함수를 convertToUserHand로 이름 변경
readUserInput함수에서 호출하던 convertToUserHand함수를 generateUserHand라는 함수를 새로 생성하여 호출
readUserInput에서 Int를 리턴하도록 한 후 converToUserHand에 전달인자로 전달

carti1108 committed:
refactor: magic literal처리, default부분 수정

computerHand부분의 randomNumber의 범위를 상수로 선언
computerHand부분의 default 부분을 .exit을 할당하는 것으로 수정

devjoon committed:
refactor: playGame 함수의 들여쓰기 수정

Hoon94 commented:
STEP1 고생하셨습니다. 다음 STEP도 화이팅입니다.🔥
